python/nx-cugraph/nx_cugraph/algorithms/dag.py

```diff
@@ -0,0 +1,55 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import cupy as cp
+import networkx as nx
+import numpy as np
+import pylibcugraph as plc
+
+from nx_cugraph.convert import _to_graph
+from nx_cugraph.utils import index_dtype, networkx_algorithm
+
+__all__ = [
+    "descendants",
+    "ancestors",
+]
+
+
+def _ancestors_and_descendants(G, source, *, is_ancestors):
+    G = _to_graph(G)
+    if source not in G:
+        hash(source)  # To raise TypeError if appropriate
```

Comment:
I was thinking about this more and realized it might be better to have a function each NX algo defined then called to verify it's inputs, then that same function could be called by backends too - no code duplication and guaranteed consistency.  This would also make it possible to support applications that call `nx-cugraph` directly.  The pattern could be something like this:
```python
######## NX dispatcher
def _dispatch(func):
    def _set_assert_args(aa_func):
        func._assert_args = aa_func
    func._arg_assertion = _set_assert_args
    ...
    return func

######## NX algo
@_dispatch
def foo(a, b):
  foo._assert_args(a, b)
  return a + b

@foo._arg_assertion
def _(a, b):
    if not(isinstance(a, int)):
        raise TypeError(f"expected int for a, got {type(a).__name__}")
    if not(isinstance(b, int)):
        raise TypeError(f"expected int for b, got {type(b).__name__}")

######## backend
def backend_foo(a, b):
    foo._assert_args(a, b)
    return a + b
```

but I agree that this would be a lot of work to refactor every dispatchable NX algo to apply this pattern everywhere.  We could do this incrementally though...
I don't think this should block this PR though, but maybe we should file an issue to NX (cc @MridulS)

Reply:
I should say, I don't think the lack of reusable algo arg assertion funcs should block the PR, but I think the update to check if `source is not None` and raise `ValueError` to match NX should be done.

Reply:
NetworkX raises `NetworkXError` for these algorithms if source is `None`, so we are correct how we match in this case. I think we may need to be more careful about `None` as a node (which isn't allowed) elsewhere--you're right, `ValueError: None cannot be a node` is somewhat common.